src/lib/coreaudio/CoreAudioDevice.cxx

```diff
@@ -0,0 +1,676 @@
+		char buffer[bufferSize];
+		
+		if (CFStringGetCString(deviceName, buffer, bufferSize, kCFStringEncodingUTF8))
+		{
+			name.assign(buffer);
+		}
+		CFRelease(deviceName);
+	}
+	return name;
+}
+
+UInt32
+CoreAudioDevice::GetTotalOutputChannels() const
+{
+	UInt32 channels = 0;
+
+	if (!device_id)
+		return channels;
+
+	AudioObjectPropertyAddress  aopa;
+	aopa.mScope    = kAudioDevicePropertyScopeOutput;
+	aopa.mElement  = 0;
+	aopa.mSelector = kAudioDevicePropertyStreamConfiguration;
+
+	UInt32 size = 0;
+	OSStatus err = AudioObjectGetPropertyDataSize(device_id, &aopa, 0, NULL, &size);
+	if (err != noErr)
+		return channels;
+
+	AudioBufferList* buffer_list = (AudioBufferList*)malloc(size);
+	err = AudioObjectGetPropertyData(device_id, &aopa, 0, NULL, &size, buffer_list);
+	if (err == noErr)
+	{
+		for(UInt32 buffer = 0; buffer < buffer_list->mNumberBuffers; ++buffer)
+			channels += buffer_list->mBuffers[buffer].mNumberChannels;
+	}
+	else
+	{
+		FormatError(macos_output_domain, "Unable to get total device output channels - id: 0x%04x. Error = %s", (uint)device_id, GetError(err).c_str());
+	}
+
+	free(buffer_list);
+
+	return channels;
+}
+
+UInt32
+CoreAudioDevice::GetNumChannelsOfStream(UInt32 stream_idx) const
+{
+	UInt32 channels = 0;
+
+	if (!device_id)
+		return channels;
+  
+	AudioObjectPropertyAddress  aopa;
+	aopa.mScope    = kAudioDevicePropertyScopeOutput;
+	aopa.mElement  = 0;
+	aopa.mSelector = kAudioDevicePropertyStreamConfiguration;
+
+	UInt32 size = 0;
+	OSStatus err = AudioObjectGetPropertyDataSize(device_id, &aopa, 0, NULL, &size);
+	if (err != noErr)
+		return channels;
+  
+	AudioBufferList* buffer_list = (AudioBufferList*)malloc(size);
+	err = AudioObjectGetPropertyData(device_id, &aopa, 0, NULL, &size, buffer_list);
+	if (err == noErr)
+	{
+		if (stream_idx < buffer_list->mNumberBuffers)
+			channels = buffer_list->mBuffers[stream_idx].mNumberChannels;
+	}
+	else
+	{
+	  FormatError(macos_output_domain, "Unable to get number of stream output channels - id: 0x%04x. Error = %s.",
+				  (uint)device_id, GetError(err).c_str());
+	}
+  
+	free(buffer_list);
+  
+	return channels;
+}
+
+bool
+CoreAudioDevice::GetStreams(AudioStreamIdList* stream_id_list)
+{
+	if (!stream_id_list || !device_id)
+		return false;
+
+	// Get only output streams (mScope)
+	AudioObjectPropertyAddress  aopa;
+	aopa.mScope    = kAudioDevicePropertyScopeOutput;
+	aopa.mElement  = 0;
+	aopa.mSelector = kAudioDevicePropertyStreams;
+
+	UInt32  size = 0;
+	OSStatus err = AudioObjectGetPropertyDataSize(device_id, &aopa, 0, NULL, &size);
+	if (err != noErr)
+		return false;
+
+	UInt32 stream_count = size / sizeof(AudioStreamID);
+	AudioStreamID* streams_list = new AudioStreamID[stream_count];
+	
+	err = AudioObjectGetPropertyData(device_id, &aopa, 0, NULL, &size, streams_list);
+	if (err == noErr)
+	{
+		for (UInt32 stream = 0; stream < stream_count; stream++)
+			stream_id_list->push_back(streams_list[stream]);
```

Comment:
Return value instead of passing an output-parameter!

Reply:
Done